opentelemetry-sdk/src/opentelemetry/sdk/_metrics/metric_reader_storage.py

```diff
@@ -12,23 +12,105 @@
 from opentelemetry.sdk._metrics.measurement import Measurement
 from opentelemetry.sdk._metrics.point import Metric
 from opentelemetry.sdk._metrics.sdk_configuration import SdkConfiguration
+from opentelemetry.sdk._metrics.view import View
+
+
+# TODO: #2300
+class ViewStorage:
+    def __init__(
+        self,
+        view: View,
+        instrument: Instrument,
+        sdk_config: SdkConfiguration,
+    ) -> None:
+        pass
+
+    def consume_measurement(self, measurement: Measurement) -> None:
+        pass
+
+    def collect(self, temporality: AggregationTemporality) -> Iterable[Metric]:
+        pass
 
 
-# TODO: #2378
 class MetricReaderStorage:
     """The SDK's storage for a given reader"""
 
     def __init__(self, sdk_config: SdkConfiguration) -> None:
-        pass
+        self._lock = RLock()
+        self._sdk_config = sdk_config
+        self._view_storages: Dict[Instrument, List[ViewStorage]] = {}
+
+    def _get_or_init_view_storage(
+        self, instrument: Instrument
+    ) -> List["ViewStorage"]:
+        # Optimistically get the relevant views for the given instrument. Once set for a given
+        # instrument, the mapping will never change
+        if instrument in self._view_storages:
+            return self._view_storages[instrument]
+
+        with self._lock:
+            # double check if it was set before we held the lock
+            if instrument in self._view_storages:
+                return self._view_storages[instrument]
+
+            # not present, hold the lock and add a new mapping
+            view_storages = []
+            for view in self._sdk_config.views:
+                if view.match(instrument):
+                    # Note: if a view matches multiple instruments, this will create a separate
+                    # ViewStorage per instrument. If the user's View configuration includes a
+                    # name, this will cause multiple conflicting output streams.
+                    view_storages.append(
+                        ViewStorage(
+                            view=view,
+                            instrument=instrument,
+                            sdk_config=self._sdk_config,
+                        )
+                    )
+
+            # if no view targeted the instrument, use the default
+            if not view_storages:
+                view_storages.append(
+                    ViewStorage(
+                        view=default_view(instrument),
+                        instrument=instrument,
+                        sdk_config=self._sdk_config,
+                    )
+                )
+            self._view_storages[instrument] = view_storages
+            return view_storages
 
     def consume_measurement(self, measurement: Measurement) -> None:
-        pass
+        storages = self._get_or_init_view_storage(measurement.instrument)
+        for storage in storages:
+            storage.consume_measurement(measurement)
```

Comment:
```suggestion
        for view_storage in self._get_or_init_view_storage(measurement.instrument)
            view_storage.consume_measurement(measurement)
```
For consistency, it is a bit hard to follow when there are many names for the symbols.